tests/test_queue.py

```diff
@@ -476,3 +484,62 @@ def test_ApiJobQueue_stop_results_in_queue_threads_not_running(mocker):
 
         job_queue.main_thread.quit.assert_called_once_with()
         job_queue.download_file_thread.quit.assert_called_once_with()
+
+
+def test_ApiJobQueue_emits_main_queue_updated_signal_when_message_or_reply_download_added(
+    mocker,
+):
+    mock_client = mocker.MagicMock()
+    mock_session_maker = mocker.MagicMock()
+
+    with threads(2) as [main_thread, file_download_thread]:
+        job_queue = ApiJobQueue(mock_client, mock_session_maker, main_thread, file_download_thread)
+
+    message_download_job = MessageDownloadJob("mock", "mock", "mock")
+    reply_download_job = ReplyDownloadJob("mock", "mock", "mock")
+    main_queue_updated_emissions = QSignalSpy(job_queue.main_queue.queue.queue_updated_signal)
+    job_queue.main_thread.isRunning = mocker.MagicMock(return_value=True)
+    job_queue.download_file_thread.isRunning = mocker.MagicMock(return_value=True)
+
+    job_queue.enqueue(message_download_job)
+    job_queue.enqueue(reply_download_job)
+    main_queue_updated_emissions.wait(MAX_SIGNAL_WAITING_TIME)
+
+    assert len(main_queue_updated_emissions) == 2
+    assert main_queue_updated_emissions[0][0] == 1
+    assert main_queue_updated_emissions[1][0] == 2
+
+    job_queue.main_queue.queue.get()
+    job_queue.main_queue.queue.get()
+    main_queue_updated_emissions.wait(MAX_SIGNAL_WAITING_TIME)
+
+    assert len(main_queue_updated_emissions) == 4
+    assert main_queue_updated_emissions[2][0] == 1
+    assert main_queue_updated_emissions[3][0] == 0
+
+
+def test_ApiJobQueue_does_not_emit_main_queue_updated_signal_when_non_message_job_added(
+    mocker,
+):
+    mock_client = mocker.MagicMock()
+    mock_session_maker = mocker.MagicMock()
+
+    with threads(2) as [main_thread, file_download_thread]:
+        job_queue = ApiJobQueue(mock_client, mock_session_maker, main_thread, file_download_thread)
+
+    message_download_job = SeenJob("mock", "mock", "mock")
```

Comment:
Some indentation is missing in this test and the one above.

```py
    with threads(2) as [main_thread, file_download_thread]:
        job_queue = ApiJobQueue(mock_client, mock_session_maker, main_thread, file_download_thread)

        # Caution: when using the threads() context manager, you need to indent the entire block,
        # because the threads are stopped when exiting the context. (That's what's actually being managed.)
        message_download_job = SeenJob("mock", "mock", "mock")
        # etc.
```

References:

- [Definition of  the `threads` context manager](https://github.com/freedomofpress/securedrop-client/blob/0.8.1/securedrop_client/app.py#L190-L204)
- It was introduced because the Qt really dislikes when `QThread` objects are destroyed before the corresponding threads are closed. (https://github.com/freedomofpress/securedrop-client/commit/5d1dd0c6b84e444822f16680b0cb1b355a44a0a9).

Reply:
Thanks very much for the feedback and extra context!

Reply:
nerdy python pun not intended 🙂